cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,114 +285,675 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+		m.degradedState.LogErrorf("cannot verify ubuntu-save mount at %v: %v",
+			boot.InitramfsUbuntuSaveDir, err)
+
+		// we are done, even if errors occurred
+		return nil, err
+	}
+
+	m.degradedState.setPartitionKeyValue("ubuntu-save", "mount-state", "mounted")
+	m.degradedState.setPartitionKeyValue("ubuntu-save", "mount-location", boot.InitramfsUbuntuSaveDir)
+
+	// all done, nothing left to try and mount
+	return nil, nil
+}
+
+func generateMountsModeRecover(mst *initramfsMountsState) error {
+	// steps 1 and 2 are shared with install mode
+	if err := generateMountsCommonInstallRecover(mst); err != nil {
 		return err
 	}
-	if !matches {
-		return fmt.Errorf("cannot validate boot: ubuntu-data mountpoint is expected to be from disk %s but is not", disk.Dev())
+
+	// get the disk that we mounted the ubuntu-seed partition from as a
+	// reference point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+	if err != nil {
+		return err
 	}
-	if haveSave {
-		// 3.2a we have ubuntu-save, verify it as well
-		matches, err = disk.MountPointIsFromDisk(boot.InitramfsUbuntuSaveDir, diskOpts)
+
+	// 3. run the state machine logic for mounting partitions, this involves
+	//    trying to unlock then mount ubuntu-data, and then unlocking and
+	//    mounting ubuntu-save
+	//    see the state* functions for details of what each step does and
+	//    possible transition points
+
+	// ensure that the last thing we do after mounting everything is to lock
+	// access to sealed keys
+	defer func() {
+		if err := secbootLockTPMSealedKeys(); err != nil {
+			logger.Noticef("error locking access to sealed keys: %v", err)
+		}
+	}()
+
+	// first state to execute is to unlock ubuntu-data with the run key
+	machine := newStateMachine(disk)
+	for {
+		final, err := machine.execute()
+		// TODO: consider whether certain errors are fatal or not
 		if err != nil {
 			return err
 		}
-		if !matches {
-			return fmt.Errorf("cannot validate boot: ubuntu-save mountpoint is expected to be from disk %s but is not", disk.Dev())
+		if final {
+			break
 		}
 	}
 
-	// 4. final step: copy the auth data and network config from
-	//    the real ubuntu-data dir to the ephemeral ubuntu-data
-	//    dir, write the modeenv to the tmpfs data, and disable
-	//    cloud-init in recover mode
-	if err := copyUbuntuDataAuth(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+	// 3.1 write out degraded.json
+	b, err := json.Marshal(machine.degradedState)
+	if err != nil {
 		return err
 	}
-	if err := copyNetworkConfig(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+
+	// needed?
+	err = os.MkdirAll(boot.InitramfsHostUbuntuDataDir, 0755)
+	if err != nil {
 		return err
 	}
-	if err := copyUbuntuDataMisc(boot.InitramfsHostUbuntuDataDir, boot.InitramfsDataDir); err != nil {
+
+	err = ioutil.WriteFile(filepath.Join(boot.InitramfsHostUbuntuDataDir, "degraded.json"), b, 0644)
```

Comment:
sorry I missed this, this might the real data volume in some cases though (if we managed to open data but not save), wondering if writing there is good idea, we might have to pick a more ephemeral place instead

Reply:
What about `dirs.SnapBootstrapRunDir` ?

Reply:
sounds good to me, @pedronis wdyt?

Reply:
sounds good